### PR TITLE
pgcraft/pipeline_utils: support mapping single uniforms to dedicated uniform buffers

### PR DESCRIPTION
**Pull request by mbouron-gpsw**

### CODE REVIEW

File: libnodegl/pgcraft.c [hunk 1]
```diff
@@ -41,6 +41,63 @@
...
+    int glsl_version;
+    const char *glsl_version_suffix;
+    const char *sym_vertex_index;
+    const char *sym_instance_index;
+    const char *rg; // 2-component texture picking (could be either rg or ra depending on the OpenGL version)
+    int has_in_out_qualifiers;
+    int has_in_out_layout_qualifiers;
+    int has_precision_qualifiers;
+    int has_modern_texture_picking;
+    int has_explicit_bindings;
```
Comment (mbouron-gpsw) [inline libnodegl/pgcraft.c:99]: NOTE: it might be a good idea to re-order some fields here.

File: libnodegl/pipeline.h [hunk 2]
```diff
@@ -22,6 +22,7 @@
 #ifndef PIPELINE_H
 #define PIPELINE_H
 
+#include "block.h"
```
Comment (cboesch-gpsw) [inline libnodegl/pipeline.h:25]: is this needed?

File: libnodegl/pipeline_utils.c [hunk 3]
```diff
@@ -19,20 +19,131 @@
...
+    int ret;
+    if ((ret = ngli_pipeline_init(s->pipeline, pipeline_params)) < 0 ||
+        (ret = ngli_pipeline_set_resources(s->pipeline, pipeline_resources)) < 0)
+        return ret;
+
+    s->compat_info = params->compat_info;
+    if (s->compat_info->use_blocks) {
+        if (!(gpu_ctx->features & NGLI_FEATURE_BUFFER_MAP)) {
+            LOG(ERROR, "context does not support mapping buffers");
+            return NGL_ERROR_GRAPHICS_UNSUPPORTED;
```
Comment (cboesch-gpsw) [inline libnodegl/pipeline_utils.c:73]: Is this user-triggerable?

File: libnodegl/pipeline_utils.c [hunk 4]
```diff
@@ -19,20 +19,131 @@
...
+        }
+
+        for (int i = 0; i < NGLI_PROGRAM_SHADER_NB; i++) {
+            const struct block *block = &s->compat_info->blocks[i];
+            if (block->size) {
```
Comment (cboesch-gpsw) [inline libnodegl/pipeline_utils.c:78]: can you reduce the depth with a `continue`?

File: libnodegl/pipeline_utils.c [hunk 5]
```diff
@@ -19,20 +19,131 @@
...
+                s->uniform_buffers[i] = buffer;
+
+                int ret = ngli_buffer_init(buffer,
+                                           block->size,
+                                           NGLI_BUFFER_USAGE_DYNAMIC_BIT |
+                                           NGLI_BUFFER_USAGE_UNIFORM_BUFFER_BIT);
+                if (ret < 0)
+                    return ret;
+
+                ret = ngli_buffer_map(buffer, buffer->size, 0, (void **)&s->mapped_datas[i]);
```
Comment (cboesch-gpsw) [inline libnodegl/pipeline_utils.c:91]: So you need a cast? Maybe the prototype should be `void *datap` then

File: libnodegl/pipeline.h (hunk 2 above)
Comment (mbouron-gpsw) [reply libnodegl/pipeline.h]: Nope, it's a leftover from when the pipeline was directly handling the compat layer.

File: libnodegl/pgcraft.c [hunk 6]
```diff
@@ -1404,6 +1404,41 @@ const struct darray *ngli_pgcraft_get_texture_infos(const struct pgcraft *s)
...
 }
 
+struct program *ngli_pgcraft_get_program(const struct pgcraft *s)
+{
+    return s->program;
+}
+
+struct pipeline_layout ngli_pgcraft_get_pipeline_layout(const struct pgcraft *s)
+{
+    struct pipeline_layout layout = {
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/pgcraft.c]: Done.

File: libnodegl/pass.c [hunk 7]
```diff
@@ -539,19 +534,26 @@ int ngli_pass_prepare(struct pass *s)
...
 
     desc->pipeline = ngli_pipeline_create(gpu_ctx);
     if (!desc->pipeline)
         return NGL_ERROR_MEMORY;
 
+    const struct pipeline_params pipeline_params = {
+        .type          = s->pipeline_type,
+        .graphics      = pipeline_graphics,
+        .program       = ngli_pgcraft_get_program(desc->crafter),
+        .layout        = ngli_pgcraft_get_pipeline_layout(desc->crafter),
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/pass.c]: Fixed.

File: libnodegl/hud.c [hunk 8]
```diff
@@ -1288,19 +1279,30 @@ int ngli_hud_init(struct hud *s)
...
     s->pipeline = ngli_pipeline_create(gpu_ctx);
     if (!s->pipeline)
         return NGL_ERROR_MEMORY;
 
+    const struct pipeline_params pipeline_params = {
+        .type          = NGLI_PIPELINE_TYPE_GRAPHICS,
+        .graphics      = {
+            .topology    = NGLI_PRIMITIVE_TOPOLOGY_TRIANGLE_STRIP,
+            .state       = graphicstate,
+            .rt_desc     = rnode->rendertarget_desc,
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/hud.c]: Fixed.

File: libnodegl/backends/gl/buffer_gl.c [hunk 9]
```diff
@@ -68,6 +68,15 @@ int ngli_buffer_gl_upload(struct buffer *s, const void *data, int size, int offs
     return 0;
 }
 
+int ngli_buffer_gl_map(struct buffer *s, int size, int offset, void **data)
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/backends/gl/buffer_gl.c]: Renamed.

File: libnodegl/pipeline_utils.c (hunk 4 above)
Comment (mbouron-gpsw) [reply libnodegl/pipeline_utils.c]: Done.

File: libnodegl/pipeline_utils.c (hunk 5 above)
Comment (mbouron-gpsw) [reply libnodegl/pipeline_utils.c]: The prototype maps the Vulkan API:
``` c
VkResult vkMapMemory(
    VkDevice                                    device,
    VkDeviceMemory                              memory,
    VkDeviceSize                                offset,
    VkDeviceSize                                size,
    VkMemoryMapFlags                            flags,
    void**                                      ppData);
```

Comment (cboesch-gpsw) [reply libnodegl/pipeline_utils.c]: `void*` is compatible with a `void**`, you won't need a cast in this direction

Comment (mbouron-gpsw) [reply libnodegl/pipeline_utils.c]: On the other hand I will need something like this for OpenGL:
``` c
*(void **)datap = glMapBuffer(...);
```

File: libnodegl/pgcraft.c [hunk 10]
```diff
@@ -1483,6 +1566,13 @@ void ngli_pgcraft_freep(struct pgcraft **sp)
     ngli_darray_reset(&s->texture_infos);
     ngli_darray_reset(&s->vert_out_vars);
 
+    struct pgcraft_compat_info *compat_info = &s->compat_info;
+    if (compat_info->use_blocks) {
```
Comment (cboesch-gpsw) [inline libnodegl/pgcraft.c:1570]: do you need the condition?

File: libnodegl/pipeline_utils.c [hunk 11]
```diff
@@ -19,20 +19,143 @@
...
+    const struct pipeline_resources *pipeline_resources = params->resources;
+
+    int ret;
+    if ((ret = ngli_pipeline_init(s->pipeline, pipeline_params)) < 0 ||
+        (ret = ngli_pipeline_set_resources(s->pipeline, pipeline_resources)) < 0)
+        return ret;
+
+    s->compat_info = params->compat_info;
+    if (s->compat_info->use_blocks) {
+        int ret = init_buffers(s, params);
```
Comment (cboesch-gpsw) [inline libnodegl/pipeline_utils.c:108]: an `int ret` already exists in this scope

File: libnodegl/pipeline_utils.c [hunk 12]
```diff
@@ -19,20 +19,143 @@
...
+        if (layout->buffers_desc[i].type == NGLI_TYPE_UNIFORM_BUFFER &&
+            layout->buffers_desc[i].stage == stage &&
+            layout->buffers_desc[i].binding == binding) {
+            return i;
+        }
+    }
+    return -1;
+}
+
+static int init_buffers(struct pipeline_compat *s, const struct pipeline_compat_params *params)
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/pipeline_utils.c]: Renamed to `init_blocks_buffers()`.

File: libnodegl/pipeline_utils.c (hunk 11 above)
Comment (mbouron-gpsw) [reply libnodegl/pipeline_utils.c]: Fixed.

File: libnodegl/pgcraft.c [hunk 13]
```diff
@@ -1021,7 +1067,8 @@ static int craft_vert(struct pgcraft *s, const struct pgcraft_params *params)
         (ret = inject_uniforms(s, b, params, NGLI_PROGRAM_SHADER_VERT)) < 0 ||
         (ret = inject_texture_infos(s, params, NGLI_PROGRAM_SHADER_VERT)) < 0 ||
         (ret = inject_blocks(s, b, params, NGLI_PROGRAM_SHADER_VERT)) < 0 ||
-        (ret = inject_attributes(s, b, params)) < 0)
+        (ret = inject_attributes(s, b, params)) < 0 ||
+        (ret = inject_ublock(s, b, NGLI_PROGRAM_SHADER_VERT)))
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/pgcraft.c]: Fixed.

File: libnodegl/pgcraft.c [hunk 14]
```diff
@@ -639,6 +657,34 @@ const char *ublock_names[] = {
...
+        /* instance name is empty to make field accesses identical to uniform accesses */
+        .instance_name = "",
+        .type          = NGLI_TYPE_UNIFORM_BUFFER,
+        .stage         = stage,
+        .block         = block,
+        .buffer        = NULL,
+    };
+    snprintf(pgcraft_block.name, sizeof(pgcraft_block.name), "ngl_%s", ublock_names[stage]);
+
+    int binding = inject_block(s, b, &pgcraft_block);
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/pgcraft.c]: Done.

File: libnodegl/pgcraft.c (hunk 10 above)
Comment (mbouron-gpsw) [reply libnodegl/pgcraft.c:1570]: Removed.

File: libnodegl/pgcraft.c [hunk 15]
```diff
@@ -1201,10 +1261,15 @@ static void probe_texture_info_elems(const struct pgcraft *s, struct pgcraft_tex
...
+        else if (is_sampler_or_image(field->type)) {
             field->index = get_texture_index(s, field->name);
-        else
-            field->index = get_uniform_index(s, field->name);
+        } else {
+            const struct pgcraft_compat_info *compat_info = &s->compat_info;
+            if (compat_info->use_blocks)
+                field->index = get_ublock_index(s, field->name, field->stage);
+            else
+                field->index = get_uniform_index(s, field->name);
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/pgcraft.c]: Done (called ngli_pgcraft_get_uniform_index())

File: libnodegl/pipeline_utils.c (hunk 5 above)
Comment (cboesch-gpsw) [reply libnodegl/pipeline_utils.c]: OK

**Review by cboesch-gpsw — APPROVED** [review]

**Pull request merged**